In [ ]:
#!pip install ast

In [1]:
import pandas as pd
import numpy as np
import ast

/anaconda3/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: numpy.dtype size changed, may indicate binary incompatibility. Expected 96, got 88
  return f(*args, **kwds)


In [2]:
content_df = pd.read_csv("/Users/shriyashekhar/Documents/IWork/Data/Output/content_nouns.csv", encoding="ISO-8859-1")
content_df.head()

StreamID                                            Content  \
0       163  TXmAk2KZAy4\r\nNMeUjebo1Ac\r\nEEuTxFhp3go\r\nC...   
1       419  TXmAk2KZAy4\r\nNMeUjebo1Ac\r\nEEuTxFhp3go\r\nC...   
2       507  wBYKUgUyGWc\r\nA team of world-class drivers, ...   
3       199  Castrol EDGE  is Castrol?s flagship power bran...   
4       201  Charles ?Cheers? Wakefield, Castrol?s founder,...   

                                       Content_Nouns  
0  [('Castrol', 'NNP'), ('oil', 'NN'), ('TITANIUM...  
1  [('Castrol', 'NNP'), ('oil', 'NN'), ('TITANIUM...  
2  [('wBYKUgUyGWc', 'NN'), ('team', 'NN'), ('worl...  
3  [('Castrol', 'NNP'), ('EDGE', 'NNP'), ('Castro...  
4  [('Charles', 'NNP'), ('\x93Cheers\x94', 'NNP')...

In [3]:
def get_nouns_for_stream(stream_id):
    stream_nouns = content_df[content_df["StreamID"] == stream_id]["Content_Nouns"].values
    #stream_nouns = stream_nouns.astype(str)
    first_stream_noun = stream_nouns[0]
#     print("Printing debug info....")
#     print(first_stream_noun)
#     print(type(first_stream_noun))
#     print("End of printing debug info....")
    
    if isinstance(first_stream_noun, str):
        # https://stackoverflow.com/questions/10775894/converting-a-string-representation-of-a-list-into-an-actual-list-object
        first_stream_noun = ast.literal_eval(first_stream_noun)
        
    return [x[0] for x in first_stream_noun]

### Nouns for the stream 507 as a list as below...

In [4]:
get_nouns_for_stream(199)

['Castrol',
 'EDGE',
 'Castrol\x92s',
 'flagship',
 'power',
 'brand',
 'PCOs',
 'performance',
 'engine',
 'oil',
 'performing',
 'vehicles',
 'OEMs',
 'world',
 'Castrol',
 'EDGE',
 'TITANIUM',
 'FST\x99',
 'oil',
 'strength',
 'oil',
 'film',
 'protection',
 'oil',
 'pressures',
 'engines',
 'today',
 'video',
 'Today',
 'engines',
 'pressure',
 'areas',
 'engine',
 'pressures',
 'excess',
 'square',
 'centimetre',
 'that\x92s',
 'cars',
 'stiletto',
 'hee',
 'video',
 'Castrol',
 'EDGE',
 'strength',
 'technology',
 'pressures',
 'high-performance',
 'engines',
 'Castrol',
 'EDGE',
 'problem',
 'engine',
 'pressures',
 'TITANIUM',
 'FST\x99',
 'Fluid',
 'Strength',
 'Technology',
 'TITANIUM',
 'FST\x99',
 'oil',
 'strength',
 'oil',
 'breakdown',
 'sha',
 'TITANIUM',
 'FST\x99',
 'adapts',
 'engine',
 'need',
 'pressures',
 'contact',
 'variety',
 'engine',
 'spee',
 'Castrol',
 'R',
 'amp',
 'D',
 'Technology',
 'world',
 'Castrol',
 'Stress',
 'Test',
 'film',
 'strength',
 'Cast

## Shriya: So you call the method like this: get_nouns_for_stream(507) to get a list of nouns for that stream with the specified stream ID. You can then loop through the items in the list and do your word net and SPARQL analysis.

In [5]:
# from nltk.corpus import wordnet as wn

# list_of_nouns = get_nouns_for_stream(201)
# print(len(list_of_nouns))

# for i in range(0, len(list_of_nouns)):
#     for j in range(0, len(list_of_nouns)):
# # for noun in list_of_nouns:
# #     print(i)
# #     if not wn.synsets(list_of_nouns[i]):
#         while(len(wn.synsets(list_of_nouns[i]))==0):
#             i+=1
#         a = wn.synsets(list_of_nouns[i])[0]
# #         i+=1
#         while(len(wn.synsets(list_of_nouns[j]))==0):
#     #         print(i)
#             j+=1
#         print(j)
#         b = wn.synsets(list_of_nouns[j])[0]
#         if a and b:
#             print('two terms')
#             print(a)
#             print(b)
#             print('parent')
#             print(a.lowest_common_hypernyms(b))


# # return a.lowest_common_hypernyms(b)

In [6]:
from nltk.corpus import wordnet as wn
words = get_nouns_for_stream(201)
syns = {w : [] for w in words}

for k, v in syns.items():
    for synset in wn.synsets(k):
        for lemma in synset.lemmas():
            v.append(lemma.name())
# print(syns)

In [7]:
dog = wn.synsets('dog')[0]
cat = wn.synsets('cat')[0]
cat.lowest_common_hypernyms(dog)

[Synset('carnivore.n.01')]

In [8]:
oil = wn.synsets('oil')[1]
castrol = wn.synsets('liquid')[0]
oil.lowest_common_hypernyms(castrol)

[Synset('substance.n.01')]

In [9]:
wn.synset('policeman.n.01').lowest_common_hypernyms(wn.synset('chef.n.01'))

[Synset('person.n.01')]

In [10]:
bata = wn.synsets('Wakefield')
if bata:
    print("empty")

In [11]:
len(wn.synsets('liquid'))

11

In [12]:
wn.synsets('liquid')[0].name().split(".")[0]

'liquid'

In [40]:
import nltk
from nltk.corpus import wordnet as wn
dog = wn.synset('petrol.n.01')
cat = wn.synset('diesel.n.01')
float(dog.path_similarity(cat))

0.1111111111111111

In [57]:
from nltk.corpus import wordnet as wn

# list_of_nouns_1 = get_nouns_for_stream(201)
# list_of_nouns_1 =['Charles', 'oil', 'Wakefield', 'power']
# # list_of_nouns_2 = get_nouns_for_stream(199)
# list_of_nouns_2 = ['Castrol', 'petrol', 'engines', 'flagship']
list_of_nouns_1 =['petrol', 'Malaysia']
# list_of_nouns_2 = get_nouns_for_stream(199)
list_of_nouns_2 = ['gasoline','Indonesia']
# print(list_of_nouns_1)
# print(list_of_nouns_2)

a_1 = []
b_1 = []
for i in range(0, len(list_of_nouns_1)):
    for j in range(0, len(list_of_nouns_2)):
# for noun in list_of_nouns:
        
#     if not wn.synsets(list_of_nouns[i]):
        if(len(wn.synsets(list_of_nouns_1[i]))==0):
            a_1.append(list_of_nouns_1[i])
            i+=1
        a = wn.synsets(list_of_nouns_1[i])[0]
        print("a")
#         i+=1
        if(len(wn.synsets(list_of_nouns_2[j]))==0):
            b_1.append(list_of_nouns_2[j])
            j+=1
#         print(j)
        b = wn.synsets(list_of_nouns_2[j])[0]
        print("b")
        if a and b:
#             print(a.lowest_common_hypernyms(b))
            x = a.path_similarity(b)
            print(a)
            print(b)
            print(x)
            print(a.lowest_common_hypernyms(b))
            if((x is not None) and (x >= 0.1) and (a is not b)):
                
#                 print('similar terms')
#                 print(a)
#                 print(b)
#                 print('hypernym')
#                 print(a.lowest_common_hypernyms(b))
                a_1.append(a.lowest_common_hypernyms(b))
                b_1.append(a.lowest_common_hypernyms(b))
            else:
                a_1.append("")
                b_1.append("")
print(a_1)
print(b_1)
                # print(b_1)

#                 print(a_1)


# return a.lowest_common_hypernyms(b)

a
b
Synset('gasoline.n.01')
Synset('gasoline.n.01')
1.0
[Synset('gasoline.n.01')]
a
b
Synset('gasoline.n.01')
Synset('indonesia.n.01')
0.08333333333333333
[Synset('physical_entity.n.01')]
a
b
Synset('malaysia.n.01')
Synset('gasoline.n.01')
0.07692307692307693
[Synset('physical_entity.n.01')]
a
b
Synset('malaysia.n.01')
Synset('indonesia.n.01')
0.25
[Synset('country.n.02')]
['', '', '', [Synset('country.n.02')]]
['', '', '', [Synset('country.n.02')]]


In [ ]:
list_of_nouns_1 =['Charles', 'EDGE', 'Wakefield', 'Castrol\x92s']
for i in range(0, 4):
    print(wn.synsets(list_of_nouns_1[i]))
    print(i)
    while(len(wn.synsets(list_of_nouns_1[i]))==0):
        print("bcg")
        i+=1
        print(i)
    

In [54]:
wn.synsets('Malaysia')

[Synset('malaysia.n.01')]

In [ ]:
a.lowest_common_hypernyms(b)